### PR TITLE
Verify blocks on import

### PR DESCRIPTION
**Pull request by mikhailOK**
Verify that hash struct field matches block contents and signature matches the hash and public keys.
TODO: support rejecting in pending_blocks logic

### CODE REVIEW

File: node/client/src/lib.rs [hunk 1]
```diff
@@ -340,6 +359,16 @@ impl Client {
...
                 .expect(BEACON_SHARD_BLOCK_MATCH);
 
+            let bls_keys = self.get_authority_keys(next_beacon_block.index());
+            if !next_beacon_block.signature.verify(&bls_keys, next_beacon_block.hash.as_ref()) ||
+                !next_shard_block.signature.verify(&bls_keys, next_shard_block.hash.as_ref()) {
+                error!(target: "client", "Importing a block with an incorrect signature ({:?}, {:?}); signers: ({:?},{:?})",
+                       next_shard_block.block_hash(), next_beacon_block.block_hash(),
+                       next_beacon_block.signature.authority_count(),
+                       next_shard_block.signature.authority_count());
+                // TODO: handle rejecting correctly
```
Comment (nearmax) [inline node/client/src/lib.rs]: Extend `BlockImportingResult` with `InvalidBlock` and return it here.

File: node/client/src/lib.rs [hunk 2]
```diff
@@ -281,6 +284,19 @@ impl Client {
...
+    /// Checks that the cached hash matches the content of the block
+    pub fn verify_block_hash(beacon_block: &SignedBeaconBlock, shard_block: &SignedShardBlock) -> bool {
+        shard_block.hash == hash_struct(&shard_block.body.header) &&
+            beacon_block.hash == hash_struct(&beacon_block.body.header) &&
+            beacon_block.body.header.shard_block_hash == shard_block.hash
+    }
+
+    /// Gets BLS keys for validating GroupSignature at block_index
+    pub fn get_authority_keys(&self, block_index: u64) -> Vec<BlsPublicKey> {
+        let (_, authority_map) = self.get_uid_to_authority_map(block_index);
```
Comment (ilblackdragon) [inline node/client/src/lib.rs]: off topic here: but @nearmax why is it a authority *map*? it has ids from 0 to num_authorities. Why is not just a Vec?

Comment (nearmax) [reply node/client/src/lib.rs]: Historical reasons from the times of continuous TxFlow and rotating authorities. Currently we are also planning on rotating authorities, but this is done in a more clean-cut manner where at each block we have a different group of authorities. While in continuous TxFlow an each group is associated with some epoch and these groups can overlap in time, meaning one group is still exiting the DAG, while another one is already building in exactly the same DAG. I do not remember exactly with Vec representation but it was problematic to express stuff like this with Vec only.

File: node/client/src/lib.rs [hunk 3]
```diff
@@ -340,17 +361,35 @@ impl Client {
...
 
+            let bls_keys = self.get_authority_keys(next_beacon_block.index());
+            if !next_beacon_block.signature.verify(&bls_keys, next_beacon_block.hash.as_ref()) ||
+                !next_shard_block.signature.verify(&bls_keys, next_shard_block.hash.as_ref()) {
+                error!(target: "client", "Importing a block with an incorrect signature ({:?}, {:?}); signers: ({:?},{:?})",
+                       next_shard_block.block_hash(), next_beacon_block.block_hash(),
+                       next_beacon_block.signature.authority_count(),
+                       next_shard_block.signature.authority_count());
+                // TODO enable when we sign blocks with second BLS
+                if false {
```
Comment (nearmax) [inline node/client/src/lib.rs]: I do not think that's where we should be verifying the signatures of the block.
Verification should be happening before block is added to `pending_*_blocks`.

**Pull request merged**